Add hyperkit version check whether user's hyperkit is newer

govargo opened the pull request:
### What type of PR is this?

/kind bug

### What this PR does / why we need it:

(Updated 2019/11/11)
I removed point1 of the PR due to duplication of #5840.

<del>#### 1. FastFail if driver error occurres.</del>

<del>So far, even if driver error occurres(like not installed) minikube keeps to starting VM when `minikube start`.</del>
<del>Of course this finish with failure.</del>
<del>This PR fix this behaviour.</del>
<del>If driver error occurres, minikube stop starting VM(minikube exits).</del>

#### 2. Add version check of hyperkit which inspect user's hyperkit is newer than specific version

If hyperkit's version is older than expected, `minikube start(hyperkit)` will fail due to version.
Related Issue #5594, #5780

This PR add suggestion about hyperkit upgrade information, if user's hyperkit version is older than expected.

### Which issue(s) this PR fixes:
Fixes #5478, #5821

### Does this PR introduce a user-facing change?
<del>Yes. This PR add hyperkit version check to `minikube start` command and FastFail if driver error occurres.</del>
Yes. This PR add hyperkit version check to `minikube start` command.

(Removed FastFail due to duplication of #5840)<details>
**Before this PR, minikube keeps to starting VM(but keep tp failing) **

```
$ minikube start --vm-driver=hyperkit
😄  minikube v1.5.1 on Darwin 10.14.5

⚠️  'hyperkit' driver reported a possible issue: exec: "hyperkit": executable file not found in $PATH
💡  Suggestion: Run 'brew install hyperkit'
📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/hyperkit/

💾  Downloading driver docker-machine-driver-hyperkit:
    > docker-machine-driver-hyperkit.sha256: 65 B / 65 B [---] 100.00% ? p/s 0s
    > docker-machine-driver-hyperkit: 10.79 MiB / 10.79 MiB  100.00% 4.57 MiB p
🔑  The 'hyperkit' driver requires elevated permissions. The following commands will be executed:

    $ sudo chown root:wheel /Users/k-iso/.minikube/bin/docker-machine-driver-hyperkit
    $ sudo chmod u+s /Users/k-iso/.minikube/bin/docker-machine-driver-hyperkit


Password:
💿  Downloading VM boot image ...
    > minikube-v1.5.1.iso.sha256: 65 B / 65 B [--------------] 100.00% ? p/s 0s
    > minikube-v1.5.1.iso: 143.76 MiB / 143.76 MiB [-] 100.00% 64.66 MiB p/s 3s
🔥  Creating hyperkit VM (CPUs=2, Memory=2000MB, Disk=20000MB) ...
🔄  Retriable failure: create: Error creating machine: Error in driver during machine creation: new-ing Hyperkit: Could not find hyperkit executable
🔥  Deleting "minikube" in hyperkit ...
🔥  Creating hyperkit VM (CPUs=2, Memory=2000MB, Disk=20000MB) ...
🔄  Retriable failure: create: Error creating machine: Error in driver during machine creation: new-ing Hyperkit: Could not find hyperkit executable
🔥  Deleting "minikube" in hyperkit ...
🔥  Creating hyperkit VM (CPUs=2, Memory=2000MB, Disk=20000MB) ...
🔄  Retriable failure: create: Error creating machine: Error in driver during machine creation: new-ing Hyperkit: Could not find hyperkit executable
🔥  Deleting "minikube" in hyperkit ...
🔥  Creating hyperkit VM (CPUs=2, Memory=2000MB, Disk=20000MB) ...
🔄  Retriable failure: create: Error creating machine: Error in driver during machine creation: new-ing Hyperkit: Could not find hyperkit executable
🔥  Deleting "minikube" in hyperkit ...

💣  Unable to start VM: create: Error creating machine: Error in driver during machine creation: new-ing Hyperkit: Could not find hyperkit executable

😿  Sorry that minikube crashed. If this was unexpected, we would love to hear from you:
👉  https://github.com/kubernetes/minikube/issues/new/choose
```

**After this PR, minikube fail fast**

```
$ minikube start --vm-driver=hyperkit
😄  minikube v1.5.1 on Darwin 10.14.5

⚠️  'hyperkit' driver reported a possible issue: exec: "hyperkit": executable file not found in $PATH
💡  Suggestion: Run 'brew install hyperkit'
📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/hyperkit/

```
</details>

**After this PR, minikube checks user's hyperkit version**

If user's hyperkit current version is 0.20180621 and specific version is 0.20190201, minikube suggests hyperkit upgrade information.

```
$ minikube start --vm-driver=hyperkit
😄  minikube v1.5.2 on Darwin 10.14.5

⚠️   'hyperkit' driver reported a possible issue: the installed hyperkit version (0.20180621) is older than the minimum recommended version (0.20191201)
💡  Suggestion: Run 'brew upgrade hyperkit'
📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/hyperkit/

💾  Downloading driver docker-machine-driver-hyperkit:
    > docker-machine-driver-hyperkit.sha256: 65 B / 65 B [---] 100.00% ? p/s 0s
    > docker-machine-driver-hyperkit: 10.79 MiB / 10.79 MiB  100.00% 6.34 MiB p
🔑  The 'hyperkit' driver requires elevated permissions. The following commands will be executed:

    $ sudo chown root:wheel /Users/k-iso/.minikube/bin/docker-machine-driver-hyperkit
    $ sudo chmod u+s /Users/k-iso/.minikube/bin/docker-machine-driver-hyperkit


Password:
💿  Downloading VM boot image ...
    > minikube-v1.5.1.iso.sha256: 65 B / 65 B [--------------] 100.00% ? p/s 0s
    > minikube-v1.5.1.iso: 143.76 MiB / 143.76 MiB [-] 100.00% 17.73 MiB p/s 8s
🔥  Creating hyperkit VM (CPUs=2, Memory=2000MB, Disk=20000MB) ...
🐳  Preparing Kubernetes v1.16.2 on Docker '18.09.9' ...
💾  Downloading kubeadm v1.16.2
💾  Downloading kubelet v1.16.2
🚜  Pulling images ...
🚀  Launching Kubernetes ...
⌛  Waiting for: apiserver
🏄  Done! kubectl is now configured to use "minikube"
```

medyagh commented on Makefile:
```diff
@@ -24,6 +24,8 @@ ISO_VERSION ?= v$(VERSION_MAJOR).$(VERSION_MINOR).1
 DEB_VERSION ?= $(subst -,~,$(RAW_VERSION))
 RPM_VERSION ?= $(DEB_VERSION)
+# used by hyperkit versionCheck(whether user's hyperkit is older than this)
+HYPERKIT_VERSION ?= 0.20190201
```
I am not sure if this is the best place to store the mininum required hyperkit version

medyagh replied:
I think a better place for this is in the package itself as a constant

govargo replied:
I moved this constant.
Makefile(& pkg/version/version.go) → pkg/minikube/registry/drvs/hyperkit.go

medyagh commented on cmd/minikube/cmd/start.go:
```diff
@@ -588,6 +588,7 @@ func selectDriver(oldConfig *cfg.Config) string {
 			out.ErrT(out.Documentation, "Documentation: {{.url}}", out.V{"url": st.Doc})
 		}
 		out.ErrLn("")
+		os.Exit(1)
```
I believe we have a wrapper for exit in Out package, please use that

govargo replied:
Originally the purpose of this PR was FastFail, but I deleted because this is overlaps with # 5840.

medyagh commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +88,57 @@ func status() registry.State {
+	// If hyperkit is not newer than specificVersion, suggest upgrade information
+	isNew := isNewerVersion(convertVersionToDate(string(out)), specificVersion)
+	if !isNew {
+		return registry.State{Installed: true, Error: fmt.Errorf("%s is too old. the higher than %s is needed", "hyperkit version", version.GetHyperKitVersion()), Fix: "Run 'brew upgrade hyperkit' or Upgrade Docker for Desktop version", Doc: docURL}
```
Lets make sure this is a non-fatal error and only warn the user ! as someone might not have corp access or not allowed to install a newer version.

medyagh replied:
the error message could be rephrased to

"the mininum recommended hyperkit version is .... your version is .... please consider upgrading your hyperkit.
Fix : brew upgrade hyperkit 
for more information read ... (link to doc on minikube site)
"

govargo replied:
I fixed this to be non-fatal error and only warn the user.

```
⚠️   'hyperkit' driver reported a possible issue: the mininum recommended hyperkit version is 0.20190201. your hyperkit version is 0.20180621. please consider upgrading your hyperkit
💡  Suggestion: Run 'brew upgrade hyperkit'
📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/hyperkit/
```

tstromberg commented on cmd/minikube/cmd/start.go:
```diff
@@ -588,6 +588,7 @@ func selectDriver(oldConfig *cfg.Config) string {
 			out.ErrT(out.Documentation, "Documentation: {{.url}}", out.V{"url": st.Doc})
 		}
 		out.ErrLn("")
+		os.Exit(1)
```
Let's collect more evidence before determining that every warning message should result in an exit. For instance: an older release of hyperkit is probably fine for many people.

I do agree for exiting early in the case of a driver that just isn't installed -- see #5840

Also, minikube should usually not force exits without consulting the `--force` variable.

medyagh replied:
I agree I recommend not exiting... We don't have enough evidence it won't work for everyone. There are some evidence reinstallling docker machine fixes it.... But overall it's good to detect super old hyperkits and let user know of it

govargo replied:
Originally the purpose of this PR was FastFail, but I deleted this line because this is overlaps with # 5840.

govargo replied:
I left only the hyperkit version check logic.

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +88,57 @@ func status() registry.State {
+	currentVersionDate, err := time.Parse(layout, currentVersion)
+	if err != nil {
+		glog.Warningf("Unable to parse to time.Date: %v", err)
+	}
```
`isNewerVersion` should probably return the error rather than silently dropping it, so that the caller can determine what the right course of action is.

govargo replied:
Thanks!
I changed that `isNewerVersion` returns error if `err != nil`.

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +88,57 @@ func status() registry.State {
+// splitHyperKitVersion splits version from v0.YYYYMMDD-HH-xxxxxxx or 0.YYYYMMDD to YYYYMMDD
+func splitHyperKitVersion(version string) string {
+	// Cut off like "20190201-11-gc0dd46" or "20190201"
+	version = strings.Split(version, ".")[1]
```
Handle the case where the version does not contain a ".": otherwise it will panic with out of range.

govargo replied:
I added handle of the case where the version does not contain a "."
And I also dded the unit test case.

tstromberg commented on pkg/version/version.go:
```diff
@@ -58,6 +61,11 @@ func GetISOPath() string {
 
+// GetHyperKitVersion returns the specific hyperKit version
+func GetHyperKitVersion() string {
+	return hyperKitVersion
```
Move the expected version into `drvs/hyperkit` - as it's only consumed there.

govargo replied:
I moved this constant and getter method.
Makefile(& pkg/version/version.go) → pkg/minikube/registry/drvs/hyperkit.go

govargo commented:
Thank you for review @tstromberg @medyagh!

I'll retake these.
And when that is done, I will return comments.

govargo commented:
@tstromberg @medyagh

I'm sorry for late.
I updated my code as you reviewed.

And FastFail(if vm-driver does not installed) will be implemented by #5840, so I removed FastFail from this PR.
I just only left hyperkit version check.

I also updated this PR title & description.

Thank you.

govargo commented:
Updated code to resolve conflict(latest codebase edited the same file)

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +92,70 @@ func status() registry.State {
+}
+
+// GetHyperKitVersion returns the specific hyperKit version
+func GetHyperKitVersion() string {
```
I think you can safely remove this public function. This variable is only apparently used by this file.

govargo replied:
You are right.
I modified this to private function.

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -37,6 +39,11 @@ const (
 
+var (
+	// hyperkitSpecificVersion is used by hyperkit versionCheck(whether user's hyperkit is older than this)
+	hyperkitSpecificVersion = "0.20190201"
```
Instead of `hyperkit.hyperkitSpecificVersion` - can this be renamed to `minimumVersion`?

govargo replied:
Thanks nice naming!
I renamed.

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +92,70 @@ func status() registry.State {
+		return registry.State{Installed: true, Error: fmt.Errorf("hyperkit version check failed:\n%s", err), Doc: docURL}
+	}
+	if !isNew {
+		return registry.State{Installed: true, Error: fmt.Errorf("the mininum recommended hyperkit version is %s. your hyperkit version is 0.%s. please consider upgrading your hyperkit", GetHyperKitVersion(), currentVersion), Fix: "Run 'brew upgrade hyperkit'", Doc: docURL}
```
set `Healthy: true`, as we don't really know yet if the hyperkit driver is healthy or not even if the version is old.

This way the hyperkit driver can still be automatically selected, but the error displayed.

govargo replied:
> This way the hyperkit driver can still be automatically selected, but the error displayed.

Thanks. I added Healthy property.

tstromberg commented:
The current text reads:

```
⚠️  'hyperkit' driver reported an issue: the mininum recommended hyperkit version is 0.20191201. your hyperkit version is 0.20190802. please consider upgrading your hyperkit
💡  Suggestion: Run 'brew upgrade hyperkit'
```

Could the first sentence be rephrased to be shorter? Perhaps:

```
⚠️  'hyperkit' driver reported an issue: the installed hyperkit version (0.20191201) is older than the minimum recommended version (0.20191201)
💡  Suggestion: Run 'brew upgrade hyperkit'
```

govargo commented:
@tstromberg 
Thank you for your good feedback!!

I modified as you reviewed.
And I changed warning message as you proposed.

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +92,70 @@ func status() registry.State {
+	// If current hyperkit is not newer than minimumVersion, suggest upgrade information
+	isNew, err := isNewerVersion(currentVersion, specificVersion)
+	if err != nil {
+		return registry.State{Installed: true, Healthy: true, Error: fmt.Errorf("hyperkit version check failed:\n%s", err), Doc: docURL}
```
use %v for err, rather than %s. Don't ask me why, it's just a Go common practice. :)

govargo replied:
Thanks. I'll remember this manner.

govargo commented:
I updated PR description.

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +92,70 @@ func status() registry.State {
+	layout := "20060102"
+	currentVersionDate, err := time.Parse(layout, currentVersion)
+	if err != nil {
+		glog.Warningf("Unable to parse to time.Date: %v", err)
```
In general we either: return errors, or log them. We almost never do  both, as it often causes duplicate errors to be logged.

In order to get the same basic effect here, use:

`return false, errors.Wrap(err, "parse date")`

govargo replied:
> In general we either: return errors, or log them. We almost never do both, as it often causes duplicate errors to be logged.

Thanks. 
I'll remember this!

tstromberg commented on pkg/minikube/registry/drvs/hyperkit/hyperkit.go:
```diff
@@ -85,5 +92,70 @@ func status() registry.State {
+	return version
+}
+
+// getHyperKitVersion returns the minimum hyperKit version
```
`Get` is implied for Go function names. This would normally just be called `hyperkit.hyperKitVersion`. I would rather just see this method removed, since it does not add anything over referring  to `minimumVersion` directly.

govargo replied:
Sorry, you are right.
I removed this getter func.

govargo commented:
Thank you for your quick feedback.
I will work on immediately

govargo commented:
@tstromberg 
I updated.

Thanks for your suggestion.
I'll remember those.

tstromberg commented:
Some test failures:

```
# k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit [k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit.test]
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:55:15: undefined: splitHyperKitVersion
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:85:15: undefined: convertVersionToDate
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:119:18: undefined: isNewerVersion
```

tstromberg commented:
/ok-to-test

govargo commented:
> Some test failures:

```
# k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit [k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit.test]
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:55:15: undefined: splitHyperKitVersion
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:85:15: undefined: convertVersionToDate
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:119:18: undefined: isNewerVersion
```

Is this test executed by master branch?
If so, the test will fail because these method don't exist yet.

https://travis-ci.org/kubernetes/minikube/jobs/611933306?utm_medium=notification&utm_source=github_status

govargo commented:
I close this PR because the minikube v1.6 do this, unless they specify --vm-driver=hyperkit.

tstromberg commented:
Re-opening because this PR is still quite valuable to us, for version checking.

tstromberg commented:
Travis still failing, but the tests run correctly when diff'd against master.  Do you mind merging your PR against master (`git merge master`) to see if it fixes the test?

```
# k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit [k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit.test]
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:55:15: undefined: splitHyperKitVersion
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:85:15: undefined: convertVersionToDate
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:119:18: undefined: isNewerVersion
```

govargo commented:
@tstromberg
Thanks for reopening this PR.

> Travis still failing, but the tests run correctly when diff'd against master. Do you mind merging your PR against master (git merge master) to see if it fixes the test?

OK, I'll rebase against latest master branch.

govargo commented:
I updated

govargo commented:
/retest

govargo commented:
The test result doesn't change...

```
# k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit [k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit.test]
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:55:15: undefined: splitHyperKitVersion
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:85:15: undefined: convertVersionToDate
pkg/minikube/registry/drvs/hyperkit/hyperkit_test.go:119:18: undefined: isNewerVersion
ok  	k8s.io/minikube/pkg/minikube/registry	0.026s	coverage: 95.1% of statements
FAIL	k8s.io/minikube/pkg/minikube/registry/drvs/hyperkit [build failed]
```

I doubt this test fails due to git or go mod, but I have no conviction. I have no idea...

If someone noticed this cause, would you tell me please.

tstromberg commented:
I don't get what is going on here. I'm just going to try to merge this and see if master works.

tstromberg commented:
PR seems to work just fine at master. ¯\_(ツ)_/¯

govargo commented:
Thank you very much!!! @tstromberg
